libcxx/include/__algorithm/ranges_clamp.h

```diff
@@ -37,9 +37,10 @@ struct __fn {
     _LIBCPP_ASSERT_UNCATEGORIZED(!bool(std::invoke(__comp, std::invoke(__proj, __high), std::invoke(__proj, __low))),
                                  "Bad bounds passed to std::ranges::clamp");
 
-    if (std::invoke(__comp, std::invoke(__proj, __value), std::invoke(__proj, __low)))
+    auto&& __projected = std::invoke(__proj, __value);
+    if (std::invoke(__comp, std::forward<decltype(__projected)>(__projected), std::invoke(__proj, __low)))
       return __low;
-    else if (std::invoke(__comp, std::invoke(__proj, __high), std::invoke(__proj, __value)))
+    else if (std::invoke(__comp, std::invoke(__proj, __high), std::forward<decltype(__projected)>(__projected))
```

Comment:
There is actually a problem here. We can potentially do a double-move. Imagine we have the following comparator:

```
struct Foo {
  std::string s;
};

// taking by value is important here
auto comparator = [](std::string a, std::string b) {
  return std::atoi(a.c_str()) < std::atoi(b.c_str());
};

auto projection = [](Foo const& foo) {
  return foo.s;
};

Foo foo{"12"};
Foo high{"10"};
Foo low{"1"};
std::ranges::clamp(foo, low, high, comparator, projection);
```

Here, we end up calling `std::invoke(comp, forward<...>(projected), low)`, it returns false and then we call `std::invoke(comp, forward<...>(projected), high)`. We end up moving `projected` twice into the `invoke` calls, so the second time we get an empty string.

Is that right? If so, this would need a test too.

Reply:
Ok, so actually it turns out that MSVC does the same thing here and they went through the exact same questioning as us, see https://github.com/microsoft/STL/pull/1898#discussion_r625118642.

However, I [played around](https://godbolt.org/z/fMM7nrn74) and I wasn't able to craft a case where passing an lvalue would fail (despite what @timsong-cpp said in that review). Unless we are able to come up with one, I think it would be reasonable to pass a lvalue.

Reply:
https://github.com/microsoft/STL/issues/3970#issuecomment-1685120958

Reply:
Ah, thanks a lot! Like I said in the other comment, @pandaninjas let's add a test for that and go back to `std::forward`. Then I think this patch will be good to go.